db/commitlog/commitlog_entry.hh

```diff
@@ -37,9 +37,35 @@ public:
     frozen_mutation&& mutation() && { return std::move(_mutation); }
 };
 
+namespace detail {
+class force_sync : public std::optional<db::timeout_clock::time_point> {
+    using opt_type = std::optional<db::timeout_clock::time_point>; 
+public:
+    constexpr force_sync(opt_type t = std::nullopt)
+        : opt_type(t)
+    {}
+    constexpr force_sync(db::timeout_clock::time_point t)
+        : opt_type(t)
+    {}
+    constexpr force_sync(bool b)
+        : opt_type(b ? yes : no)
+    {}
+
+    bool operator==(const force_sync& v) const {
+        return std::equal_to<opt_type>{}(*this, v);
+    }
+    bool operator!=(const force_sync& v) const {
+        return !(*this == v);
+    }
+
+    static const force_sync no;
+    static const force_sync yes;
```

Comment:
I don't think we need the optional, it just adds a lot of branches. We can use time_point::min() to force immediate sync, time_point::max() not to force any time-based sync, and a point in time to force sync at that point. Moreover we can use the timer everywhere with that, we don't have to check everywhere for min/max.

Reply:
How does it add branches? You would have the same amount of branches to check "is min", "is max" etc. It is just semantics.
And we can't very well use the timer with "max" (as in arm it at), because that is not what we meant... Comparing to next trigger time, yes.

Reply:
I mean not having checks for is_min/is_max at all.

Why can't we arm the timer at time_point::max()? It arms a time_point that will never be reached (at least while we're alive), so it's the same as no_timeout. Similarly a timer at time_point::min() will expire immediately (or at least, at the next poll period).

Reply:
Because that is not what "max" (or min, or timeout) means in the context. It means "not before, but respect periodicity". 
So either way we at the very least need to do a `std::min(timeout, active_timeout)`.